### PR TITLE
Add SSO credentials provider from AWS SDK v2

### LINKED ISSUE #2295: SSO authentication not working
## Bug report 

AWS recently added a new method of authentication: https://aws.amazon.com/single-sign-on/. However almost none of the methods AWS provides to authenticate work with Nextflow

### Expected behavior and actual behavior

* If you login using the SSO URL in your browser, you are provided with 3 options. When using "Option 1: Set AWS environment variables", I get `The AWS Access Key Id you provided does not exist in our records. (Service: Amazon S3; Status Code: 403; Error Code: InvalidAccessKeyId; Request ID: XXXXXXXXXXXXXX; S3 Extended Request ID: XXXXXXXXXXXXXXXX)`
* When using "Option 2: Add a profile to your AWS credentials file", everything works as it should, I think because this actually sets the `aws_access_key_id` and `aws_secret_access_key` in the credentials file
* Using `aws configure sso`, which creates a kind of special AWS profile that does not have an access key or id in this form, but instead has the field `sso_start_url`. Here nextflow gives me `Missing AWS security credentials -- Provide access/security keys pair or define a IAM instance profile (suggested)`
* `nextflow` also doesn't let you provide the credentials on the command line for some reason I don't understand. e.g. `nextflow run foo.nf -aws.accessKey XXXXXXXX -aws.secretKey YYYYYYYYY`. This gives `Unknown option: -aws.accessKey`.

### Steps to reproduce the problem

* Ensure you have a nextflow pipeline configured for AWS Batch
* Setup SSO: https://console.aws.amazon.com/singlesignon/home
* Create a new User
* Create a new Permission Set that is allowed to run batch jobs and edit S3 files
* Assign the new user to the permission set
* Log in as that user using either the browser or the CLI, and then follow the 3 above options

### Program output 

See above.

### Environment 

* Nextflow version: `21.04.0`
* Java version: `1.8.0_152-release-1056-b12`
* Operating system: Ubuntu Linux 21.04
* Bash version: `5.1.4`

### PR DESCRIPTION
**Pull request by bentsherman**
Close #2295 

I currently can't test it properly because we don't use SSO internally, but we will try to set up a test environment.

@multimeric feel free to give it a try in the meantime.

### CODE REVIEW

**Comment by bentsherman** [comment]
Okay, I got it to work with our test SSO user.

Run `aws configure sso` and provide the SSO start URL and SSO region. It will create a profile in `~/.aws/config` that looks like this:
```toml
[profile my-sso-profile]
sso_session = 'my-sso-session'
sso_account_id = '...'
sso_role_name = '...'
[sso-session my-sso-session]
sso_start_url = '...'
sso_region = '...'
sso_registration_scopes = 'sso:account:access'
```

You can then run a Nextflow pipeline with `aws.profile = 'my-sso-profile'` in your Nextflow configuration.

You can also run `aws configure sso --profile default` to make it the default profile.

The SSO region is just the region where the SSO user is authenticated. It is separate from the AWS region, so you can use the SSO profile and then specify any other region via `aws.region`.

**Comment by pditommaso** [comment]
Excellent. Let's add a section in the docs after "[security credentials](https://www.nextflow.io/docs/latest/aws.html#aws-security-credentials)" about this

**Comment by bentsherman** [comment]
Yes, it is documented there.

Also, tests are passing but this "snyk" test is failing, however I can't access it.

File: docs/aws.md [hunk 1]
```diff
@@ -6,21 +6,34 @@
...
-- `AWS_DEFAULT_REGION`
+1. Nextflow configuration file - `aws.accessKey` and `aws.secretKey`. See {ref}`AWS configuration<config-aws>` for more details.
 
-If `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY` are not defined in the environment, Nextflow will attempt to
-retrieve credentials from your `~/.aws/credentials` and `~/.aws/config` files. The `default` profile can be
-overridden via the environmental variable `AWS_PROFILE` (or `AWS_DEFAULT_PROFILE`).
+2. A custom profile in `~/.aws/credentials` and/or `~/.aws/config`. The profile can be supplied from the `aws.profile` config option, or the `AWS_PROFILE` or `AWS_DEFAULT_PROFILE` environmental variables.
 
-Alternatively AWS credentials and profile can be specified in the Nextflow configuration file. See {ref}`AWS configuration<config-aws>` for more details.
+3. Environment variables - `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN`.
```
Comment (pditommaso) [inline docs/aws.md:15]:
```suggestion
3. Environment variables - `AWS_ACCESS_KEY_ID` and `AWS_SECRET_ACCESS_KEY`.
```

File: docs/aws.md [hunk 2]
```diff
@@ -6,21 +6,34 @@
...
 
-:::{note}
-Credentials can also be provided by using an IAM Instance Role. The benefit of this approach is that it spares you from managing/distributing AWS keys explicitly. Read the [IAM Roles](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html) documentation and [this blog post](https://aws.amazon.com/blogs/security/granting-permission-to-launch-ec2-instances-with-iam-roles-passrole-permission/) for more details.
-:::
+4. Java System Properties - `aws.accessKeyId`, `aws.secretKey`, and `aws.sessionToken`.
+
+5. Web Identity Token credentials.
```
Comment (pditommaso) [inline docs/aws.md:19]: Not sure about this

File: docs/aws.md [hunk 3]
```diff
@@ -6,21 +6,34 @@
...
-overridden via the environmental variable `AWS_PROFILE` (or `AWS_DEFAULT_PROFILE`).
+2. A custom profile in `~/.aws/credentials` and/or `~/.aws/config`. The profile can be supplied from the `aws.profile` config option, or the `AWS_PROFILE` or `AWS_DEFAULT_PROFILE` environmental variables.
 
-Alternatively AWS credentials and profile can be specified in the Nextflow configuration file. See {ref}`AWS configuration<config-aws>` for more details.
+3. Environment variables - `AWS_ACCESS_KEY_ID`, `AWS_SECRET_ACCESS_KEY`, and `AWS_SESSION_TOKEN`.
 
-:::{note}
-Credentials can also be provided by using an IAM Instance Role. The benefit of this approach is that it spares you from managing/distributing AWS keys explicitly. Read the [IAM Roles](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/iam-roles-for-amazon-ec2.html) documentation and [this blog post](https://aws.amazon.com/blogs/security/granting-permission-to-launch-ec2-instances-with-iam-roles-passrole-permission/) for more details.
-:::
+4. Java System Properties - `aws.accessKeyId`, `aws.secretKey`, and `aws.sessionToken`.
```
Comment (pditommaso) [inline docs/aws.md:17]: This is supported indirectly but I don't think it makes sense to mention in the docs

**Comment by bentsherman** [comment]
@pditommaso regarding your comments, I have no idea if these auth methods work or are practical with Nextflow
- `AWS_SESSION_TOKEN`
- Java system properties
- Web identity token

I documented them because they are implemented in Nextflow, but I have no attachment to them. Ideally we should either document them or remove them.

**Comment by bentsherman** [comment]
I think AWS_SESSION_TOKEN and web identity token are both related to temporary credentials, so probably not practical for Nextflow. Java system properties should work fine but you might as well use environment variables or Nextflow config.

So I'm fine with not documenting these but then we should also remove them from the code 😄

**Comment by pditommaso** [comment]
The rationale is to document what's expected to be a feature. Some features can be indirectly implemented by the inherited library, but they should not be documented if there are not meant to be exposed as a supported feature.

**Pull request merged**